core/src/main/java/io/questdb/griffin/SqlCompiler.java

```diff
@@ -1611,6 +1668,95 @@ private void validateTableModelAndCreateTypeCast(
+      }
+    
+	private CompiledQuery sqlTableBackup(SqlExecutionContext executionContext) throws SqlException {
+		ObjList<CharSequence> tableNames = new ObjList<>();
+		while (true) {
+			CharSequence tok = SqlUtil.fetchNext(lexer);
+			if (null == tok) {
+				throw SqlException.position(lexer.getPosition()).put(" expected a table name");
+			}
+			CharSequence tableName = GenericLexer.unquote(tok);
+			int status = engine.getStatus(executionContext.getCairoSecurityContext(), path, tableName, 0, tableName.length());
+			if (status != TableUtils.TABLE_EXISTS) {
+				throw SqlException.position(lexer.getPosition()).put(" '").put(tableName).put("' is not  a valid table");
+			}
+			tableNames.add(tableName);
+
+			tok = SqlUtil.fetchNext(lexer);
+			if (null == tok || Chars.equals(tok, ';')) {
+				break;
+			}
+			if (!Chars.equals(tok, ',')) {
+				throw SqlException.position(lexer.getPosition()).put(" expected ','");
+			}
+		}
+		
+		for (int n=0; n<tableNames.size(); n++) {
+			backupTable(tableNames.get(n), executionContext);
+		}
+
+		return compiledQuery.ofRenameTable();
+	}
+    
+	public void backupTable(@NotNull CharSequence tableName, @NotNull SqlExecutionContext executionContext) {
+		LOG.info().$("Starting backup of ").$(tableName).$();
+		if (null == configuration.getBackupRoot()) {
+			throw CairoException.instance(0).put("Backup is disabled, no backup root directory is configured in the server configuration ['cairo.sql.backup.root' property]");
+		}
+
+		CairoSecurityContext securityContext = executionContext.getCairoSecurityContext();
+		try (TableReader reader = engine.getReader(securityContext, tableName)) {
+			cloneMetaData(tableName, configuration.getRoot(), configuration.getBackupRoot(), configuration.getMkDirMode(), (TableReaderMetadata) reader.getMetadata());
+			try (TableWriter backupWriter = engine.getBackupWriter(securityContext, tableName); RecordCursor cursor = reader.getCursor()) {
+				copyTableData(cursor, reader.getMetadata(), backupWriter);
+				backupWriter.commit();
+			}
+		}
+		LOG.info().$("Completed backup of ").$(tableName).$();
+	}
+
+	private void cloneMetaData(CharSequence tableName, CharSequence root, CharSequence backupRoot, int mkDirMode, TableReaderMetadata sourceMetaData) {
+		path.of(backupRoot).concat(tableName).put(Files.SEPARATOR).$();
+
+		if (ff.exists(path)) {
+			throw CairoException.instance(0).put("Backup dir for table \"" + tableName + "\" already exists [dir=").put(path).put(']');
+		}
+
+		if (ff.mkdirs(path, mkDirMode) != 0) {
+			throw CairoException.instance(ff.errno()).put("Could not create [dir=").put(path).put(']');
+		}
+
+		int rootLen = path.length();
+		try {
+			mem.of(ff, path.trimTo(rootLen).concat(TableUtils.META_FILE_NAME).$(), ff.getPageSize());
+			sourceMetaData.cloneTo(mem);
+	
+			// create symbol maps
+			path.trimTo(rootLen).$();
+			int symbolMapCount = 0;
+			for (int i = 0; i < sourceMetaData.getColumnCount(); i++) {
```

Comment:
this is an ok loop, but not the fastest. Loops around our codebase consistently use the following syntax:

```java
for (int i = 0, n = obj.size(); i < n; i++) {
}
```